database/memory/db.go

```diff
@@ -535,34 +587,200 @@ func (db *MemoryDB) removeAllIndices(address types.Address) error {
+			holderMap[k.Holder] = true
+		}
+	}
+	var holderArr []types.Address
+	for holdr := range holderMap {
+		holderArr = append(holderArr, holdr)
+	}
+	return holderArr, nil
 }
 
 func (db *MemoryDB) RecordERC721Token(contract types.Address, holder types.Address, block uint64, tokenId *big.Int) error {
+	db.mux.Lock()
+	defer db.mux.Unlock()
+	//find old entry
+	existingTokenEntry, errExisting := db.ERC721TokenByTokenID(contract, block-1, tokenId)
+	if errExisting != nil && errExisting != database.ErrNotFound {
+		return errExisting
+	}
+
+	//add new entry
+	tokenHolderInfo :=
+		types.ERC721Token{
+			Contract:  contract,
+			Holder:    holder,
+			Token:     tokenId.String(),
+			HeldFrom:  block,
+			HeldUntil: nil,
+		}
+	db.erc721BalancesDB = append(db.erc721BalancesDB, tokenHolderInfo)
+	/////
+	if errExisting == database.ErrNotFound {
+		return nil
+	}
+
+	blk := block - 1
+	existingTokenEntry.HeldUntil = &blk
 	return nil
 }
 
-func (db *MemoryDB) ERC721TokenByTokenID(contract types.Address, block uint64, tokenId *big.Int) (types.ERC721Token, error) {
-	return types.ERC721Token{}, database.ErrNotImplemented
+func (db *MemoryDB) ERC721TokenByTokenID(contract types.Address, block uint64, tokenId *big.Int) (*types.ERC721Token, error) {
+	db.mux.RLock()
+	defer db.mux.RUnlock()
+	var tmpItem int
+	found := false
+	for i, item := range db.erc721BalancesDB {
+		if item.Contract == contract && item.HeldFrom <= block && item.Token == tokenId.String() {
+			if !found {
+				tmpItem = i
+				found = true
+			} else {
+				if item.HeldFrom > db.erc721BalancesDB[tmpItem].HeldFrom {
+					tmpItem = i
+				}
+			}
+
+		}
+	}
+	if !found {
+		return nil, database.ErrNotFound
+	}
+	return &db.erc721BalancesDB[tmpItem], nil
 }
 
 func (db *MemoryDB) ERC721TokensForAccountAtBlock(contract types.Address, holder types.Address, block uint64, options *types.TokenQueryOptions) ([]types.ERC721Token, error) {
-	return nil, database.ErrNotImplemented
+	return db.erc721TokensAtBlock(contract, &holder, block, options)
+}
+
+func (db *MemoryDB) erc721TokensAtBlock(contract types.Address, holder *types.Address, block uint64, options *types.TokenQueryOptions) ([]types.ERC721Token, error) {
+	db.mux.RLock()
+	defer db.mux.RUnlock()
+	startTokenId := big.NewInt(-1)
+	if options.After != "" {
+		parsed, success := new(big.Int).SetString(options.After, 10)
+		if !success {
+			return nil, errors.New(`could not parse "after" token ID`)
+		}
+		startTokenId = parsed
+	}
+	tokenId := new(big.Int).Add(startTokenId, big.NewInt(1))
+
+	var tmpItem types.ERC721Token
+	found := false
+	var result []types.ERC721Token
+	for _, k := range db.erc721BalancesDB {
+		if k.Contract == contract && (holder == nil || *holder == k.Holder) && k.HeldFrom <= block {
+			ercTokenId, success := new(big.Int).SetString(k.Token, 10)
+			if !success {
+				return nil, errors.New(`could not parse "erc721" token ID`)
+			}
+			if ercTokenId.Uint64() >= tokenId.Uint64() {
```

Comment:
```suggestion
			if ercTokenId.Cmp(tokenId) >= 0 {
```

The token ID may be much larger than a `uint64`, so do a comparison on the big.Int instead.